File: pylib/qlat/rng_state.py
import cqlat as c

class RngState:

    def __init__(self, x = None, y = None):
        if x is None:
            self.cdata = c.mk_rng()
        elif isinstance(x, RngState):
            if y is None:
                # make a copy of x
                self.cdata = c.mk_rng(x)
            else:
                seed = str(y)
                self.cdata = c.mk_rng(x, seed)
        else:
            assert y is None
            # seed a new rng
            seed = str(x)
            self.cdata = c.mk_rng(rng_state_root, seed)

    def __del__(self):
        assert isinstance(self.cdata, int)
        c.free_rng(self)

    def __imatmul__(self, v1):
        assert isinstance(v1, RngState)
        c.set_rng(self, v1)
        return self

    def copy(self, is_copying_data = True):
        x = RngState()
        if is_copying_data:
            x @= self
        return x

    def split(self, seed):
        rng = RngState(self, seed)
        return rng

    def rand_gen(self):
        return c.rand_gen(self)

    def u_rand_gen(self, upper = 1.0, lower = 0.0):
        return c.u_rand_gen(self, upper, lower)

    def g_rand_gen(self, center = 0.0, sigma = 1.0):
        return c.g_rand_gen(self, center, sigma)

    def c_rand_gen(self, size):
        # size can be total_site of the lattice
        return c.c_rand_gen(self, size)

    def select(self, l):
        ri = self.rand_gen() % len(l)
        return l[ri]

rng_state_root = RngState()

File: qlat-utils/pylib/cqlat_utils/rng-state.cpp
#include "lib.h"

EXPORT(mk_rng, {
  using namespace qlat;
  PyObject* p_rng = NULL;
  PyObject* p_seed = NULL;
  if (!PyArg_ParseTuple(args, "|OO", &p_rng, &p_seed)) {
    return NULL;
  }
  RngState* prng_new = NULL;
  if (NULL == p_rng) {
    prng_new = new RngState();
  } else {
    RngState& rng = py_convert_type<RngState>(p_rng);
    if (p_seed == NULL) {
      prng_new = new RngState(rng);
    } else {
      const std::string seed = py_convert_data<std::string>(p_seed);
      prng_new = new RngState(rng, seed);
    }
  }
  return py_convert((void*)prng_new);
})

EXPORT(free_rng, {
  using namespace qlat;
  return free_obj<RngState>(args);
})

EXPORT(set_rng, {
  using namespace qlat;
  return set_obj<RngState>(args);
})

EXPORT(rand_gen, {
  using namespace qlat;
  PyObject* p_rng = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_rng)) {
    return NULL;
  }
  RngState& rng = py_convert_type<RngState>(p_rng);
  const uint64_t x = rand_gen(rng);
  return py_convert(x);
})

EXPORT(u_rand_gen, {
  using namespace qlat;
  PyObject* p_rng = NULL;
  double upper = 1.0;
  double lower = 0.0;
  if (!PyArg_ParseTuple(args, "O|dd", &p_rng, &upper, &lower)) {
    return NULL;
  }
  RngState& rng = py_convert_type<RngState>(p_rng);
  const double x = u_rand_gen(rng, upper, lower);
  return py_convert(x);
})

EXPORT(g_rand_gen, {
  using namespace qlat;
  PyObject* p_rng = NULL;
  double center = 0.0;
  double sigma = 1.0;
  if (!PyArg_ParseTuple(args, "O|dd", &p_rng, &center, &sigma)) {
    return NULL;
  }
  RngState& rng = py_convert_type<RngState>(p_rng);
  const double x = g_rand_gen(rng, center, sigma);
  return py_convert(x);
})
